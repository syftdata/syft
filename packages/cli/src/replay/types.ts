export interface TSSourcePath {
  fileName: string;
}
export interface EventSource {
  name: string;
  uiEventType: string;
  source: TSSourcePath;
  owner?: string;
  ownerSource?: TSSourcePath;
}
