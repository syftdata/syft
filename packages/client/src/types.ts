import type { ApiEvent } from './monitor/types';
import type { MetricsDependency } from './metrics_dependency_registry';

export enum SyftEventType {
  TRACK,
  PAGE,
  SCREEN,
  IDENTIFY,
  GROUP_IDENTIFY
}

export enum SyftEventTrackStatus {
  TRACKED, // event is modeled and instrumented using syft.
  SEMI_TRACKED, // event is modeled, but the library is not being used.
  NOT_TRACKED // event is not modeled.
}

export enum SyftEventValidStatus {
  VALID, // event data is valid.
  NOT_VALID, // event data is not valid.
  UNKNOWN // we don't know if the event data is valid.
}

export enum SyftEventInstrumentStatus {
  INSTRUMENTED, // event is instrumented.
  NOT_INSTRUMENTED, // event is not instrumented.
  NOT_INSTRUMENTED_VERBOSE // event is not instrumented, and includes a lot of events.
}

export enum NamingCase {
  KEEP_AS_IS,
  SNAKE, // Replace spaces with underscore and lowercase.
  CAMEL, // Remove spaces/underscores and convert first letter to uppercase.
  PASCAL,
  TITLE,
  LOWER,
  UPPER
}

export interface SyftProps {
  eventName: string;
  eventType: SyftEventType;
  isValid: boolean;
}

export interface SyftEvent extends Record<string, any> {
  syft: SyftProps;
}

export const SyftEnv = {
  Prod: 'prod',
  Dev: 'dev'
} as const;

export type SyftEnvType = typeof SyftEnv;
export type SyftEnvValueType = SyftEnvType[keyof SyftEnvType];

// config provided during code generation
export interface StaticConfig {
  projectName: string;
  version: string;
  strict?: boolean; // drops events if they fail invalid checks.
  eventNameCase?: NamingCase;
  propertyNameCase?: NamingCase;
  lintingDisabled?: boolean;
}

export interface MonitorConfig {
  batchSize?: number;
  batchFlushSeconds?: number;
  remote?: string | ((events: ApiEvent[]) => Promise<boolean>);
  samplingRate?: number;
  disabled?: boolean;
}

// config provided during syft runtime initialization.
export interface RuntimeConfig {
  env?: SyftEnvValueType;
  appVersion: string;
  plugins: ISyftPlugin[];
  verbose?: boolean;
  monitor?: MonitorConfig;
  apiKey?: string;
}

export interface InternalConfig {
  monitor: Required<MonitorConfig>;
}

export type FullConfig = Required<StaticConfig> &
  Required<RuntimeConfig> &
  InternalConfig;

export const DEFAULT_STATIC_CONFIG: Required<StaticConfig> = {
  version: '0.0.0',
  strict: false,
  projectName: '',
  eventNameCase: NamingCase.TITLE,
  propertyNameCase: NamingCase.KEEP_AS_IS,
  lintingDisabled: false
};
export const DEFAULT_RUNTIME_CONFIG: Required<RuntimeConfig> = {
  env: SyftEnv.Prod,
  appVersion: '0.0.0',
  plugins: [],
  verbose: false,
  monitor: {},
  apiKey: ''
};

export const DEFAULT_MONITOR_CONFIG: Required<MonitorConfig> = {
  batchSize: 30,
  batchFlushSeconds: 30,
  samplingRate: 0.01, // 1% of the data
  // remote: 'https://events.syftdata.com/reflect/v1',
  remote: '',
  disabled: false
};

export interface IReflector {
  reflectEvent: (
    eventName: string,
    eventProperties: Record<string, any>
  ) => void;
}

export interface ISyftPlugin extends MetricsDependency {
  /**
   * Gives you an opportunity to start loading the plugin dependencies
   */
  load?: () => void;
  /**
   * Return true when the plugin loaded its dependencies
   * @returns true when the plugin is ready
   */
  isLoaded: () => boolean;

  /**
   * Initialize your plugin with the reflector.
   * This is needed in order to get benefit out of syft's extensive tooling.
   * @param reflector
   */
  init: (reflector: IReflector) => void;

  /** Gets called when an event is logged.
   * A plugin can use `eventName` and `eventType` properties to route the event.
   */
  logEvent: (event: SyftEvent) => boolean;

  /** Gets called when syft.resetUser() is called. */
  resetUserProperties: () => void;
}

export interface MonitorResponse {
  error?: Error;
  stopUntil?: number; // number of milli seconds to disable.
  // to disable until the next day, set it to 24 * 60 * 60 * 1000
}

export interface IMonitor {
  monitor: (
    events: ApiEvent[],
    onCompleted: (response: MonitorResponse) => void
  ) => void;
}

export interface IJsonUploader {
  upload: (url: string, jsonData: string) => Promise<string>;
}

export interface IConfigStore {
  set: (key: string, value: any) => void;
  get: (key: string) => any;
  remove: (key: string) => void;
}

export interface ExternalEvent {
  type: string;
  detail: any;
}
export interface IEventDispatcher {
  dispatch: (event: ExternalEvent) => any;
}
