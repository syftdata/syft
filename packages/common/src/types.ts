/* eslint-disable @typescript-eslint/no-empty-interface */
import { type StaticConfig, type SyftEventType } from './client_types';
import { type Expression } from 'ts-morph';

interface BasicInfo {
  name: string;
  documentation?: string;
}

/**
 * All supported types.
 * string
 * number
 * boolean
 * int
 * float
 * Object
 * Array
 * int
 * CUID
 * CUID2
 * UUID
 * Email
 * Url
 * CountryCode
 * Date
 * Array
 * Object
 */

export interface CUID2 {}
export interface CUID {}
export interface UUID {}
export interface Email {}
export interface Url {}
export interface CountryCode {}

export interface TypeField extends BasicInfo {
  type: TypeSchema;
  isOptional: boolean;
}

export interface TypeSchema extends BasicInfo {
  typeFields?: TypeField[]; // fields if the type is a complex object. for primitives, this will be empty.
  zodType: string; // for zod validation. z.string()
  syfttype?: string; // types defined in syft system. UUID, email etc.,
}

export interface Field extends TypeField {
  defaultValue?: string; // eg: default values 1, "hello". makes the field optional.
}

export interface EventSchema extends TypeSchema {
  fields: Field[];
  traits?: string[];
  destinations?: string[];

  eventType: SyftEventType;
  exported?: boolean;
}

export interface AST {
  eventSchemas: EventSchema[];
  syftConfig?: Expression;
  config: StaticConfig;
}
