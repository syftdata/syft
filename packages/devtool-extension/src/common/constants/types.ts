export type Size = "xSmall" | "small" | "medium" | "large" | "xLarge";
export type VariableType = "array" | "int" | "string" | "bool" | "double";
