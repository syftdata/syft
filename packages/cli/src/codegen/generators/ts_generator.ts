import { CodeBlockWriter, ModuleKind, Project, ScriptTarget } from 'ts-morph';
import { type AST, type TypeSchema } from '@syftdata/common/lib/types';
import {
  capitalize,
  logDetail,
  logFatal,
  logInfo,
  logVerbose,
  lowerize
} from '../../utils';
import { SyftEventType } from '@syftdata/common/lib/client_types';
import * as fs from 'fs';
import * as path from 'path';
import * as handlebars from 'handlebars';
import {
  GENERATED_TS_TEMPLATE,
  LOG_VALIDATE_METHODS_TEMPLATE
} from '../templates/ts_generated_template';
import { type PackageJson } from '../../config/pkg';
import { SYFT_DOCUMENTATION } from '../../utils/constants';

// TODO: Memoize
function getGeneratedTS(): HandlebarsTemplateDelegate<any> {
  return handlebars.compile(GENERATED_TS_TEMPLATE, { noEscape: true });
}

// TODO: Memoize
function getLogMethod(): HandlebarsTemplateDelegate<any> {
  return handlebars.compile(LOG_VALIDATE_METHODS_TEMPLATE, { noEscape: true });
}

export function generateSource(ast: AST, writer: CodeBlockWriter): void {
  const logMethodWriter = new CodeBlockWriter();
  ast.eventSchemas.forEach((schema) => {
    if (schema.exported === false) return;
    const requiredFields = schema.fields.filter(
      (field) => !field.isOptional && field.defaultValue === undefined
    );
    logMethodWriter.write(
      getLogMethod()({
        ...schema,
        isOptionalInput: requiredFields.length === 0,
        eventType: SyftEventType[schema.eventType],
        lowerName: lowerize(schema.name),
        capitalName: capitalize(schema.name)
      })
    );
  });

  const externalInterfaceDefs = ast.eventSchemas.flatMap((schema) => {
    return schema.fields.map((field) => {
      if ((field.type.typeFields?.length ?? 0) > 0) {
        return field.type;
      }
      return undefined;
    });
  });
  const externalInterfaceDefMap = externalInterfaceDefs.reduce((map, def) => {
    if (def != null) {
      map.set(def.name, def);
    }
    return map;
  }, new Map<string, TypeSchema>());

  const syftClassStr = getGeneratedTS()({
    syftConfig: ast.syftConfig?.getFullText() ?? '{}',
    interface_defs: externalInterfaceDefMap.values(), // get all interfaces used by schemas.
    schemas: ast.eventSchemas,
    logMethods: logMethodWriter.toString()
  });
  writer.write(syftClassStr);
}

export function generate(
  ast: AST,
  destinationDir: string,
  pkg: PackageJson
): void {
  const project = new Project({
    compilerOptions: {
      target: ScriptTarget.ES2015,
      declaration: true,
      declarationMap: true,
      sourceMap: false,
      strictNullChecks: true,
      module: ModuleKind.CommonJS,
      noImplicitAny: false,
      removeComments: false
    }
  });
  const dir = project.createDirectory(destinationDir);
  const sourceFile = project.createSourceFile(
    path.join(destinationDir, 'lib', 'generated.ts'),
    (writer) => {
      generateSource(ast, writer);
    },
    { overwrite: true }
  );

  logVerbose('Generated Client TS file: ');
  logVerbose(sourceFile.getText());
  const relativePath = path.relative('.', destinationDir);
  logDetail(
    `Generating Syft client (${pkg.json.version as string}) to ${relativePath}`
  );
  if (process.env.NODE_ENV !== 'test') {
    const emitResult = project.emitSync();
    logVerbose(
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      `Emit results are: ${emitResult.getDiagnostics()} ${emitResult.getEmitSkipped()}`
    );
    if (!fs.existsSync(dir.getPath())) {
      logFatal(`Failed to create the output directory at ${destinationDir}`);
      return;
    }
    logInfo(
      `:sparkles: You can now start using Syft Client in your code.\nRefer: ${SYFT_DOCUMENTATION}`
    );
    logDetail(`
\`\`\`
import Syft, { AmplitudePlugin } from '@syftdata/client'
...
const syft = new Syft({
  appVersion: "1.0.0",
  plugins: [new AmplitudePlugin(amplitude)]
})
syft.pageViewed({name: 'index'});
\`\`\``);
  } else {
    logVerbose('Skipping js generation to keep unit-tests faster');
  }
}
