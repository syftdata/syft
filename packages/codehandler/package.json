{
  "name": "@syftdata/codehandler",
  "version": "0.1.33-alpha.4",
  "description": "Code handler for Syft cli, which can serialize/deserialize internal models into typescript",
  "keywords": [
    "syft",
    "client",
    "events",
    "data",
    "instrumentation",
    "analytics"
  ],
  "author": "Nageswara Rao Mannem <nageswara@syftdata.com>",
  "homepage": "https://syftdata.com",
  "license": "Apache-2.0",
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "directories": {
    "src": "./src",
    "test": "./tests"
  },
  "files": [
    "./lib",
    "README.md",
    "LICENSE"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/syftdata/syft.git"
  },
  "bugs": {
    "url": "https://github.com/syftdata/syft/issues"
  },
  "scripts": {
    "build": "tsc",
    "test": "jest",
    "lint": "eslint --ext .ts src tests --fix",
    "prepublish": "npm run build",
    "prettier-format": "prettier --config .prettierrc 'src/**/*.ts' --write"
  },
  "devDependencies": {
    "@types/jest": "^29.4.0",
    "@types/node": "^18.13.0",
    "@typescript-eslint/eslint-plugin": "^5.51.0",
    "@typescript-eslint/parser": "^5.51.0",
    "eslint": "^8.36.0",
    "eslint-config-prettier": "^8.6.0",
    "eslint-config-standard-with-typescript": "^34.0.0",
    "eslint-plugin-import": "^2.27.5",
    "eslint-plugin-n": "^15.6.1",
    "eslint-plugin-prettier": "^4.2.1",
    "eslint-plugin-promise": "^6.1.1",
    "jest": "^29.4.2",
    "prettier": "^2.8.4",
    "ts-jest": "^29.0.5",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@es-joy/jsdoccomment": "^0.37.0",
    "@syftdata/common": "^0.1.33-alpha.4",
    "handlebars": "^4.7.7",
    "ts-morph": "^17.0.1"
  }
}
