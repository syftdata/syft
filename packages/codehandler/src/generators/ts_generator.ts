import { CodeBlockWriter } from 'ts-morph';
import {
  type EventSchema,
  type AST,
  type TypeSchema
} from '@syftdata/common/lib/types';
import { capitalize, lowerize } from '@syftdata/common/lib/utils';
import { SyftEventType } from '@syftdata/common/lib/client_types';
import * as handlebars from 'handlebars';
import {
  EXAMPLE_CALL_TEMPLATE,
  GENERATED_TS_TEMPLATE,
  LOG_VALIDATE_METHODS_TEMPLATE
} from '../templates/ts_generated_template';

// TODO: Memoize
function getGeneratedTS(): HandlebarsTemplateDelegate<any> {
  return handlebars.compile(GENERATED_TS_TEMPLATE, { noEscape: true });
}

// TODO: Memoize
function getLogMethod(): HandlebarsTemplateDelegate<any> {
  return handlebars.compile(LOG_VALIDATE_METHODS_TEMPLATE, { noEscape: true });
}

// TODO: Memoize
function getExampleUsage(): HandlebarsTemplateDelegate<any> {
  return handlebars.compile(EXAMPLE_CALL_TEMPLATE, { noEscape: true });
}

export function generateSource(ast: AST): string {
  const logMethodWriter = new CodeBlockWriter();
  ast.eventSchemas.forEach((schema) => {
    if (schema.exported === false) return;
    const requiredFields = schema.fields.filter(
      (field) => !field.isOptional && field.defaultValue === undefined
    );
    logMethodWriter.write(
      getLogMethod()({
        ...schema,
        isOptionalInput: requiredFields.length === 0,
        eventType: SyftEventType[schema.eventType],
        lowerName: lowerize(schema.name),
        capitalName: capitalize(schema.name)
      })
    );
  });

  const externalInterfaceDefs = ast.eventSchemas.flatMap((schema) => {
    return schema.fields.map((field) => {
      if ((field.type.typeFields?.length ?? 0) > 0) {
        return field.type;
      }
      return undefined;
    });
  });
  const externalInterfaceDefMap = externalInterfaceDefs.reduce((map, def) => {
    if (def != null) {
      map.set(def.name, def);
    }
    return map;
  }, new Map<string, TypeSchema>());

  const syftClassStr = getGeneratedTS()({
    syftConfig: ast.syftConfig?.getFullText() ?? '{}',
    interface_defs: externalInterfaceDefMap.values(), // get all interfaces used by schemas.
    schemas: ast.eventSchemas,
    logMethods: logMethodWriter.toString()
  });
  return syftClassStr;
}

export function generateExampleCalls(schemas: EventSchema[]): string[] {
  const templ = getExampleUsage();
  return schemas.map((schema) => {
    return templ({
      ...schema,
      lowerName: lowerize(schema.name),
      capitalName: capitalize(schema.name)
    });
  });
}
