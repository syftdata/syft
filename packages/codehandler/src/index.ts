export * from './serializer/ts_model';
export * from './deserializer/ts_model';
export * from './generators/ts_generator';
export * from './deserializer/zod_utils';
