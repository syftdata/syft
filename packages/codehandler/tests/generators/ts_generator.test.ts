/* eslint-disable @typescript-eslint/no-unused-vars */
import { SyftEventType } from '@syftdata/client';
import { generateSource } from '../../src/generators/ts_generator';
import { type Field, type AST } from '@syftdata/common/lib/types';
import { getZodTypeForSchema } from '@syftdata/codehandler';

const TEST_FIELDS: Field[] = [
  {
    name: 'backgrounded',
    type: {
      name: 'boolean',
      zodType: 'z.boolean()'
    },
    isOptional: false
  },
  {
    name: 'has_focus',
    type: {
      name: 'boolean',
      zodType: 'z.boolean()'
    },
    defaultValue: 'false',
    isOptional: true
  }
];

const TEST_ZOD_TYPE: string = getZodTypeForSchema(TEST_FIELDS);

describe('generate', () => {
  it('with ts parses an empty schema file', async () => {
    const ast: AST = {
      eventSchemas: [],
      config: {
        projectName: 'test',
        version: '1.0.0'
      }
    };
    expect(generateSource(ast)).toMatchSnapshot();
  });

  it('with ts parses an event that is not exported', async () => {
    const ast: AST = {
      eventSchemas: [
        {
          name: 'TestEvent',
          fields: TEST_FIELDS,
          traits: [],
          eventType: SyftEventType.PAGE,
          exported: false,
          zodType: TEST_ZOD_TYPE
        }
      ],
      config: {
        projectName: 'test',
        version: '1.0.0'
      }
    };
    expect(generateSource(ast)).toMatchSnapshot();
  });

  it('with ts parses an event that is exported', async () => {
    const ast: AST = {
      eventSchemas: [
        {
          name: 'TestEvent',
          fields: TEST_FIELDS,
          traits: [],
          eventType: SyftEventType.PAGE,
          exported: true,
          zodType: TEST_ZOD_TYPE
        }
      ],
      config: {
        projectName: 'test',
        version: '1.0.0'
      }
    };
    expect(generateSource(ast)).toMatchSnapshot();
  });
});
