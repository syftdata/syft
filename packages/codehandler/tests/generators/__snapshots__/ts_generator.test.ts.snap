// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate with ts parses an empty schema file 1`] = `
"
import type { StaticConfig, RuntimeConfig, SyftEvent, SyftPlugin } from './types'
import { SyftEventType, NamingCase } from './types'
import BaseSyft from './base'
import { z } from 'zod'




export default class Syft extends BaseSyft {    
  private static readonly staticConfig: StaticConfig = {};

  constructor(config: RuntimeConfig) {
    super(Syft.staticConfig, config);
    console.debug('Syft is running via generated library!');
  }

  resetUser(): void {
    this.batcher.resetUser();
  }

  
}
"
`;

exports[`generate with ts parses an event that is exported 1`] = `
"
import type { StaticConfig, RuntimeConfig, SyftEvent, SyftPlugin } from './types'
import { SyftEventType, NamingCase } from './types'
import BaseSyft from './base'
import { z } from 'zod'



/**
 * 
 */
export interface TestEvent {
    /**
     * 
     */
    backgrounded: boolean;
    /**
     * 
     */
    has_focus?: boolean;
}

const TestEventZod = z.object({
    backgrounded: z.boolean(),
    has_focus: z.boolean(),
});


export default class Syft extends BaseSyft {    
  private static readonly staticConfig: StaticConfig = {};

  constructor(config: RuntimeConfig) {
    super(Syft.staticConfig, config);
    console.debug('Syft is running via generated library!');
  }

  resetUser(): void {
    this.batcher.resetUser();
  }

  

  isValidTestEvent(event: TestEvent): boolean {
    const result = TestEventZod.safeParse(event)
    if (!result.success && this.config.verbose) {
      console.warn("TestEvent is not valid. errors: ", (result as any).error.toString())
    }
    return result.success
  }

  /**
   * 
   * @param event object with fields defined in the syft model file. 
   */
  testEvent(event: TestEvent): boolean {
    const evt = event ?? {}

    if (evt.has_focus === undefined) {
      evt.has_focus = false
    }

    const syft = { eventName: "TestEvent", eventType: SyftEventType.PAGE, isValid: true }    
    if (!this.isValidTestEvent(evt)) {
      if (this.config.strict) {
        return false
      }
      syft.isValid = false
    }
    return this.batcher.logEvent({...evt, syft})
}


}
"
`;

exports[`generate with ts parses an event that is not exported 1`] = `
"
import type { StaticConfig, RuntimeConfig, SyftEvent, SyftPlugin } from './types'
import { SyftEventType, NamingCase } from './types'
import BaseSyft from './base'
import { z } from 'zod'



/**
 * 
 */
export interface TestEvent {
    /**
     * 
     */
    backgrounded: boolean;
    /**
     * 
     */
    has_focus?: boolean;
}

const TestEventZod = z.object({
    backgrounded: z.boolean(),
    has_focus: z.boolean(),
});


export default class Syft extends BaseSyft {    
  private static readonly staticConfig: StaticConfig = {};

  constructor(config: RuntimeConfig) {
    super(Syft.staticConfig, config);
    console.debug('Syft is running via generated library!');
  }

  resetUser(): void {
    this.batcher.resetUser();
  }

  
}
"
`;
